File: TreasureData/TDClient.m
//
//  TDClient.m
//  TreasureData
//
//  Created by Mitsunori Komatsu on 12/15/14.
//  Copyright (c) 2014 Treasure Data Inc. All rights reserved.
//

// TODO: make this an internal class

#import <UIKit/UIKit.h>
#import <CommonCrypto/CommonDigest.h>
#import "TDClient.h"
@import GZIP;

static NSString *version = @"0.9.0";

@implementation TDClient

// Deprecated
- (id)initWithApiKey:(NSString *)apiKey apiEndpoint:(NSString*)apiEndpoint {
    return [self __initWithApiKey:apiKey apiEndpoint:apiEndpoint];
}

- (id)__initWithApiKey:(NSString *)apiKey apiEndpoint:(NSString*)apiEndpoint {
    NSString *projectId = [NSString stringWithFormat:@"_td %@", [self sha256Hash:apiKey]];
    self = [self initWithProjectId:projectId andWriteKey:@"dummy_write_key" andReadKey:@"dummy_read_key"];
    self.apiKey = apiKey;
    self.apiEndpoint = apiEndpoint;
    self.globalPropertiesBlock = ^NSDictionary *(NSString *eventCollection) {
        if (!NSClassFromString(@"NSUUID")) {
            return @{};
        }
        return @{@"#UUID": [[NSUUID UUID] UUIDString]};
    };
    /*
     > 5.times.inject(0){|a, i| puts a; x = 4 * (2 ** i); a += x; a}
     0
     4
     12
     28
     60
    */
    self.uploadRetryIntervalCoeficient = 4;
    self.uploadRetryIntervalBase = 2;
    self.uploadRetryCount = 5;
    self.enableRetryUploading = true;
    _session = [NSURLSession sharedSession];
    return self;
}

- (NSString *)sha256Hash:(NSString *)input {
    // Generate a cryptographic digest using the secure SHA-256 algorithm
    const char* cString = [input UTF8String];
    unsigned char hashedString[CC_SHA256_DIGEST_LENGTH];
    CC_SHA256(cString, (CC_LONG) strlen(cString), hashedString);

    // Convert the digest to an NSString hex-string for straightforward use
    NSMutableString *nsString = [NSMutableString stringWithCapacity:CC_SHA256_DIGEST_LENGTH * 2];
    for (int i = 0; i < CC_SHA256_DIGEST_LENGTH; i++) {
        [nsString appendFormat:@"%02x", hashedString[i]];
    }

    return nsString;
}

- (void)sendEvents:(NSData *)data database:(NSString *)database table:(NSString *)table completionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandler {
    NSString *urlString = [NSString stringWithFormat:@"%@/%@/%@", self.apiEndpoint, database, table];
    KCLog(@"Sending events to: %@", urlString);
    NSURL *url = [NSURL URLWithString:urlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:[NSString stringWithFormat: @"TD1 %@", self.apiKey] forHTTPHeaderField:@"Authorization"];
    [request setValue:@"application/vnd.treasuredata.v1.js+json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:@"application/vnd.treasuredata.v1.js+json" forHTTPHeaderField:@"Accept"];
    [request setValue:[NSString stringWithFormat:@"TD-iOS-SDK/%@ (%@ %@)", version, [[UIDevice currentDevice] systemName], [[UIDevice currentDevice] systemVersion]] forHTTPHeaderField:@"User-Agent"];
    
    if (_enableEventCompression) {
        [request setValue:@"gzip" forHTTPHeaderField:@"Content-Encoding"];
        [request setHTTPBody:[data gzippedData]];
    } else {
        [request setHTTPBody:data];
    }
    
    [self __sendHTTPRequest:request retryCounter:0 completionHandler:completionHandler];
}

- (void)sendHTTPRequest:(NSURLRequest *)request
            retryCounter:(int)retryCounter
       completionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandler {
    [self __sendHTTPRequest:request retryCounter:retryCounter completionHandler:completionHandler];
}

- (void)__sendHTTPRequest:(NSURLRequest *)request
            retryCounter:(int)retryCounter
       completionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandler {

    NSURLSessionDataTask *dataTask = [_session
                                      dataTaskWithRequest:request
                                      completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*)response;
        if (data) {
            completionHandler(data, response, error);
        }
        else {
            KCLog(@"dataTaskWithRequest error occurred(%@/%@)",
                  [NSNumber numberWithInt:retryCounter],
                  [NSNumber numberWithInt:self.uploadRetryCount]);
            KCLog(@"response=%@", httpResponse);

            if (!self.enableRetryUploading || retryCounter >= self.uploadRetryCount - 1) {
                // Give up retry
                completionHandler(data, response, error);
            }
            else {
                double wait = self.uploadRetryIntervalCoeficient * pow(self.uploadRetryIntervalBase, retryCounter);
                [NSThread sleepForTimeInterval:wait];
                [self __sendHTTPRequest: request
                           retryCounter: (retryCounter + 1)
                      completionHandler: completionHandler
                ];
            }
        }
    }];
    [dataTask resume];
}

- (void)__enableEventCompression:(BOOL)flag {
    _enableEventCompression = flag;
}

- (NSURLSession *)__session {
    return _session;
}

- (void)__setSession:(NSURLSession *)session {
    _session = session;
}

@end

File: TreasureData/TDClient.swift
//
//  TDClient.swift
//  TreasureData
//
//  The HTTP sender for the event engine, owned by `SwiftEventEngine`: builds a
//  Treasure Data ingest request (TD auth, content types, gzip) and drives the
//  retry loop.
//

import Foundation
import GZIP
#if canImport(UIKit)
import UIKit
#endif
import CommonCrypto

final class TDClient {

    private static let sdkVersion = "2.0.0"

    var apiKey: String = ""
    var apiEndpoint: String = ""

    var isTrackingIP: Bool = false
    var isEventCompressionEnabled: Bool = false

    var retryEnabled: Bool = true
    // Wait before next retry = intervalCoefficient * intervalBase ^ retryCount.
    var retryIntervalCoefficient: Int = 4
    var retryIntervalBase: Int = 2
    var retryCount: Int = 5

    /// Injectable for testing; defaults to the shared session.
    var uploadSession: URLSession = .shared

    init(apiKey: String, apiEndpoint: String) {
        self.apiKey = apiKey
        self.apiEndpoint = apiEndpoint
    }

    /// The on-disk buffer namespace derived from the api key; the engine reads
    /// it to scope its store. The "_td <sha256(apiKey)>" scheme is a legacy
    /// format preserved so upgrading apps reuse their existing buffer.
    var projectIdForBuffer: String { "_td \(TDClient.sha256Hash(apiKey))" }

    private static func sha256Hash(_ input: String) -> String {
        let data = Data(input.utf8)
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(buffer.count), &digest)
        }
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Build the TD ingest request for one `database`.`table` and send it,
    /// driving the retry loop.
    func sendEvents(_ data: Data,
                    database: String,
                    table: String,
                    completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
        let urlString = "\(apiEndpoint)/\(database)/\(table)"
        guard let url = URL(string: urlString) else {
            completionHandler(nil, nil, nil)
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("TD1 \(apiKey)", forHTTPHeaderField: "Authorization")
        let contentType = isTrackingIP
            ? "application/vnd.treasuredata.v1.mobile+json"
            : "application/vnd.treasuredata.v1+json"
        // Match the ObjC: Content-Type is set twice (json then the TD vnd type).
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        request.setValue(contentType, forHTTPHeaderField: "Accept")
        let systemName: String
        let systemVersion: String
        #if canImport(UIKit)
        systemName = UIDevice.current.systemName
        systemVersion = UIDevice.current.systemVersion
        #else
        systemName = "iOS"
        systemVersion = ""
        #endif
        request.setValue("TD-iOS-SDK/\(TDClient.sdkVersion) (\(systemName) \(systemVersion))",
                         forHTTPHeaderField: "User-Agent")

        if isEventCompressionEnabled {
            request.setValue("gzip", forHTTPHeaderField: "Content-Encoding")
            request.httpBody = (data as NSData).gzipped()
        } else {
            request.httpBody = data
        }

        sendHTTPRequest(request, retryCounter: 0, completionHandler: completionHandler)
    }

    private func sendHTTPRequest(_ request: URLRequest,
                                 retryCounter: Int,
                                 completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
        let task = uploadSession.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }
            if data != nil {
                completionHandler(data, response, error)
            } else {
                if !self.retryEnabled || retryCounter >= self.retryCount - 1 {
                    // Give up retry.
                    completionHandler(data, response, error)
                } else {
                    let wait = Double(self.retryIntervalCoefficient) * pow(Double(self.retryIntervalBase), Double(retryCounter))
                    Thread.sleep(forTimeInterval: wait)
                    self.sendHTTPRequest(request, retryCounter: retryCounter + 1, completionHandler: completionHandler)
                }
            }
        }
        task.resume()
    }
}
